### PR TITLE
Add feature flag for opq

### LINKED ISSUE #985: Put `OPQ` module after a feature flag, so that by default we dont need to build openblas
# Problem Statement

Only the opq module uses `blas` now, and it causes significant difficulty for us to maintain windows build.

# Desired Behavior

Make `opq` an optional feature, so most of the users dont need to build `openblas` by default:

Dependencies that can make optional in this process:

* `cblas`
* `lapack`
* `openblas-src`
* `acclerate-src`

### PR DESCRIPTION
**Pull request by TevinWang**
- Splits linalg.rs into `linalg/matrix.rs` and `linalg/svd.rs`
- Adds ff `opq` to both lance and pylance 

Closes #985

### CODE REVIEW

File: rust/src/index/vector/ivf.rs [hunk 1]
```diff
@@ -584,6 +588,7 @@ pub async fn build_ivf_pq_index(
         Ivf::new(centroids.clone())
     } else {
         // Pre-transforms
+        #[cfg(feature = "opq")]
```
Comment (chebbyChefNEQ) [inline rust/src/index/vector/ivf.rs:591]: Let’s at a block where when opq feature is not installed, generate an error if uses passes use_opq=true

File: rust/src/arrow/linalg.rs [hunk 2]
```diff
@@ -287,6 +292,7 @@ impl TryFrom<&FixedSizeListArray> for MatrixView {
 /// Single Value Decomposition.
 ///
 /// <https://en.wikipedia.org/wiki/Singular_value_decomposition>
+#[cfg(feature = "opq")]
```
Comment (eddyxu) [inline rust/src/arrow/linalg.rs:295]: Maybe just move svd to a separate module / file?

File: rust/src/arrow/linalg.rs [hunk 3]
```diff
@@ -177,6 +181,7 @@ impl MatrixView {
     }
 
     /// Dot multiply
+    #[cfg(feature = "opq")]
```
Comment (eddyxu) [inline rust/src/arrow/linalg.rs:184]: let's file a ticket to implement dot w/o Blas?

File: rust/src/index/vector.rs [hunk 4]
```diff
@@ -381,8 +383,10 @@ pub(crate) async fn open_index(
         dataset.manifest.clone(),
         100_usize,
     ));
+
     for stg in vec_idx.stages.iter().rev() {
         match stg.stage.as_ref() {
+            #[cfg(feature = "opq")]
```
Comment (eddyxu) [inline rust/src/index/vector.rs:389]: This gate might be too wide.

File: rust/src/arrow/linalg.rs (hunk 3 above)
Comment (TevinWang) [reply rust/src/arrow/linalg.rs:184]: Should I move this into its own module outside of `linalg.rs` as well?

File: python/Cargo.toml [hunk 5]
```diff
@@ -27,3 +27,6 @@ uuid = "1.3.0"
 
 [build-dependencies]
 prost-build = "0.11"
+
+[features]
+opq = ["lance/opq"]
```
Comment (wjones127) [inline python/Cargo.toml:32]: Let's keep this on by default.

```suggestion
[features]
default = ["opq"]
opq = ["lance/opq"]
```

File: rust/Cargo.toml
Comment (wjones127) [inline rust/Cargo.toml:1]: Could we also modify

```rust
[package.metadata.docs.rs]
rustc-args = ["--cfg", "docsrs"]
```

To:

```rust
[package.metadata.docs.rs]
features = []
no-default-features = true
```

In theory, this should allow our Rust docs to build.

File: rust/src/arrow/linalg/svd.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+use super::matrix::MatrixView;
+use arrow_array::Float32Array;
+
+#[allow(unused_imports)]
+#[cfg(target_os = "macos")]
+use accelerate_src;
+
+#[allow(unused_imports)]
+#[cfg(all(any(target_os = "linux", target_os = "windows"), not(docsrs)))]
```
Comment (wjones127) [inline rust/src/arrow/linalg/svd.rs:11]: The `docsrs` flag is now redundant.

```suggestion
#[cfg(any(target_os = "linux", target_os = "windows")))]
```

File: rust/src/index/vector/ivf.rs (hunk 1 above)
Comment (wjones127) [reply rust/src/index/vector/ivf.rs:591]: It looks like this was added to Python, but not yet Rust? Could you add that error to the rust function here as well?

**Review by wjones127 — CHANGES_REQUESTED** [review]
Looking good. We can remove the `docsrs` configuration. That was put there to disable opq for the sake of building the docs, so these changes make it redundant. I've added suggested changes for that.

**Review by wjones127 — APPROVED** [review]
Looks good pending CI

**Pull request merged**